File: src/main/java/net/shotbow/interestingfish/listeners/FishListener.java
package net.shotbow.interestingfish.listeners;

import net.shotbow.interestingfish.FishInfoFactory;
import net.shotbow.interestingfish.config.InterestingConfig;
import net.shotbow.interestingfish.objects.FishInfo;
import net.shotbow.interestingfish.utility.ItemUtility;
import org.bukkit.Tag;
import org.bukkit.entity.Item;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerFishEvent;
import org.bukkit.inventory.ItemStack;

import java.text.DecimalFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.Random;

/**
 * User: Austin
 * Date: 5/18/13
 * Time: 1:34 AM
 * (c) lazertester
 */
public class FishListener implements Listener {

    private final SimpleDateFormat dateFormat;
    private final DecimalFormat decimalFormat = new DecimalFormat("0.00");
    private final FishInfoFactory fishInfoFactory;
    private final InterestingConfig config;
    private final Random random = new Random();

    public FishListener(FishInfoFactory fishInfoFactory, InterestingConfig config) {
        this.fishInfoFactory = fishInfoFactory;
        this.config = config;
        dateFormat = new SimpleDateFormat(config.dateFormat);
    }

    @EventHandler
    public void onFish(PlayerFishEvent e) {
        if (e.getCaught() instanceof Item && Tag.ITEMS_FISHES.isTagged(((Item) e.getCaught()).getItemStack().getType())) {
            if (!e.getPlayer().hasPermission("interestingfish.catch"))
                return;
            if (config.requireLuckEnchant && !ItemUtility.fishingRodHasLuck(e.getPlayer().getInventory()) ||
                    config.excludeLuckEnchant && ItemUtility.fishingRodHasLuck(e.getPlayer().getInventory()))
                return;
            if (random.nextInt(100) < config.percentBreedsChance) {
                final String name = config.caughtByNickname ? e.getPlayer().getDisplayName() : e.getPlayer().getName();
                final FishInfo fishInfo = fishInfoFactory.makeNewFishInfo();
                final ItemStack fish = ((Item) e.getCaught()).getItemStack();
                ItemUtility.renameItem(fish, fishInfo.getName());
                ItemUtility.setLore(fish, new ArrayList<String>() {{
                    add(fishInfoFactory.parseColors(config.weightLabel + decimalFormat.format(fishInfo.getWeight()) + config.weightUnit));
                    add(fishInfoFactory.parseColors(config.caughtByLabel + name));
                    add(fishInfoFactory.parseColors(config.dateLabel + dateFormat.format(new Date())));
                }});
            }
        }
    }
}

File: src/main/kotlin/net/shotbow/interestingfish/listeners/FishListener.kt
package net.shotbow.interestingfish.listeners

import net.shotbow.interestingfish.FishInfoFactory
import net.shotbow.interestingfish.FishInfoFactory.Companion.withColors
import net.shotbow.interestingfish.config.InterestingConfig
import net.shotbow.interestingfish.util.ItemUtility
import net.shotbow.interestingfish.util.ItemUtility.isFish
import net.shotbow.interestingfish.util.ItemUtility.meta
import org.bukkit.entity.Item
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerFishEvent
import java.text.DecimalFormat
import java.text.SimpleDateFormat
import java.time.OffsetDateTime
import java.util.Random

class FishListener(
    private val fishInfoFactory: FishInfoFactory,
    private val config: InterestingConfig
) : Listener {
    private val dateFormat: SimpleDateFormat = SimpleDateFormat(config.dateFormat)
    private val decimalFormat = DecimalFormat("0.00")
    private val random = Random()

    @EventHandler
    fun onFish(e: PlayerFishEvent) {
        val caught = e.caught
        if (caught is Item && caught.itemStack.isFish()) {
            if (
                !e.player.hasPermission("interestingfish.catch") ||
                config.requireLuckEnchant && !ItemUtility.fishingRodHasLuck(e.player.inventory) ||
                config.excludeLuckEnchant && ItemUtility.fishingRodHasLuck(e.player.inventory)
            ) {
                return
            }

            if (random.nextInt(100) < config.percentBreedsChance) {
                val name = if (config.caughtByNickname) e.player.displayName else e.player.name
                val (name1, weight) = fishInfoFactory.makeNewFishInfo()
                val fish = caught.itemStack

                fish.meta {
                    displayName = name1
                    lore = listOf(
                        (config.weightLabel + decimalFormat.format(weight) + config.weightUnit).withColors(),
                        (config.caughtByLabel + name).withColors(),
                        (config.dateLabel + dateFormat.format(OffsetDateTime.now())).withColors()
                    )
                }
            }
        }
    }
}
